File: backend/src/services/ai.service.ts
import { isFallbackMode } from '../db/db';

export interface IAiAnalysisResult {
  category: 'Important' | 'Work' | 'Personal' | 'Promotions' | 'Social' | 'Spam' | 'Finance' | 'Updates' | 'Newsletters';
  priority: 'High' | 'Medium' | 'Low';
  summary: string;
  keyPoints: string[];
  sentiment: string;
  tasks: { text: string; dueDate?: string }[];
  phishingAnalysis: {
    riskScore: number;
    warnings: string[];
    isPhishing: boolean;
    explanation: string;
  };
}

// ==========================================
// GEMINI REST API / SDK INTEGRATION
// ==========================================
async function callGemini(prompt: string, jsonMode = false): Promise<string> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey || apiKey === 'YOUR_GEMINI_API_KEY') {
    throw new Error('Gemini API key is not configured');
  }

  try {
    const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${apiKey}`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        contents: [{
          parts: [{ text: prompt }]
        }],
        generationConfig: jsonMode ? {
          responseMimeType: 'application/json'
        } : undefined
      })
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData?.error?.message || `API error: ${response.statusText}`);
    }

    const data = await response.json();
    const textResult = data.candidates?.[0]?.content?.parts?.[0]?.text;
    if (!textResult) {
      throw new Error('Empty response from Gemini API');
    }
    return textResult;
  } catch (error: any) {
    console.error('Gemini API request failed:', error.message);
    throw error;
  }
}

// ==========================================
// SEMANTIC KEYWORD-BASED MOCK AI ENGINE
// ==========================================
function mockAnalyzeEmail(subject: string, body: string, fromName: string, fromEmail: string): IAiAnalysisResult {
  const content = `${subject} ${body}`.toLowerCase();
  const emailLower = fromEmail.toLowerCase();
  const senderLower = fromName.toLowerCase();

  let category: IAiAnalysisResult['category'] = 'Personal';
  let priority: IAiAnalysisResult['priority'] = 'Low';
  let sentiment = 'Casual / Neutral';
  const warnings: string[] = [];
  const tasks: IAiAnalysisResult['tasks'] = [];
  let riskScore = 0;
  let isPhishing = false;
  let explanation = 'Legitimate personal correspondence.';

  // 1. Phishing & Spam Detection
  if (
    content.includes('verify your account') ||
    content.includes('urgent action required') ||
    content.includes('password reset link') ||
    content.includes('suspicious login') ||
    content.includes('bank wire transfer') ||
    content.includes('inheritance') ||
    content.includes('lottery winner') ||
    content.includes('crypto double') ||
    emailLower.includes('security-verify') ||
    emailLower.includes('support-update') ||
    (content.includes('sign in') && content.includes('immediately') && !emailLower.includes('google.com') && !emailLower.includes('github.com'))
  ) {
    category = 'Spam';
    priority = 'High';
    riskScore = Math.floor(Math.random() * 30) + 70; // 70-99
    isPhishing = true;
    explanation = 'Identified phishing indicators: High urgency language demanding login verification, mismatch sender domain, and links asking for credentials.';
    warnings.push('Urgent account sign-in demands detected.');
    warnings.push('Sender email addresses resemble high-profile domains but are unofficial.');
    warnings.push('Malicious links looking like standard portal logins found.');
  } else if (
    content.includes('offer') ||
    content.includes('buy now') ||
    content.includes('discount') ||
    content.includes('coupon') ||
    content.includes('unsubscribe') ||
    content.includes('deal') ||
    content.includes('clearance')
  ) {
    category = 'Promotions';
    riskScore = Math.floor(Math.random() * 20) + 15; // 15-35
    explanation = 'Promotional email with marketing offers or shopping discounts.';
  } else if (
    content.includes('social') ||
    content.includes('linkedin') ||
    content.includes('twitter') ||
    content.includes('facebook') ||
    content.includes('instagram') ||
    content.includes('friend request') ||
    content.includes('commented on your post')
  ) {
    category = 'Social';
    explanation = 'Notification regarding activity on social network sites.';
  } else if (
    content.includes('invoice') ||
    content.includes('billing') ||
    content.includes('receipt') ||
    content.includes('payment') ||
    content.includes('credit card statement') ||
    content.includes('tax document') ||
    content.includes('bank transaction')
  ) {
    category = 'Finance';
    priority = 'Medium';
    explanation = 'Financial statement, invoice notification, or billing confirmation.';
  } else if (
    content.includes('newsletter') ||
    content.includes('weekly digest') ||
    content.includes('blog post') ||
    content.includes('reading list')
  ) {
    category = 'Newsletters';
    explanation = 'Regular subscription newsletter or informational feed.';
  } else if (
    content.includes('meeting') ||
    content.includes('calendar') ||
    content.includes('schedule') ||
    content.includes('zoom link') ||
    content.includes('google meet') ||
    content.includes('sync up') ||
    content.includes('jira') ||
    content.includes('standup') ||
    content.includes('pull request') ||
    content.includes('code review')
  ) {
    category = 'Work';
    priority = 'Medium';
    sentiment = 'Professional / Collaborative';
    explanation = 'Work-related schedule update, sync coordination, or collaborative code activity.';
  }

  // 2. High Priority overrides
  if (
    content.includes('urgent') ||
    content.includes('asap') ||
    content.includes('deadline') ||
    content.includes('important notice') ||
    content.includes('critical') ||
    senderLower.includes('boss') ||
    senderLower.includes('ceo') ||
    senderLower.includes('director') ||
    senderLower.includes('manager')
  ) {
    priority = 'High';
    sentiment = 'Urgent / Important';
  }

  // 3. Summarization & Bullet Points
  let summary = `Email from ${fromName} concerning '${subject}'.`;
  const keyPoints: string[] = [];

  if (category === 'Spam') {
    summary = `Alert: Highly suspicious phishing attempt mimicking a formal notification.`;
    keyPoints.push('Demands immediate action under the guise of an account lockdown.');
    keyPoints.push('Contains deceptive login links designed to compromise account credentials.');
    keyPoints.push('Do NOT open any attachments, links, or reply to this message.');
  } else if (category === 'Work') {
    summary = `Team sync and coordination updates regarding project items.`;
    keyPoints.push(`Sender ${fromName} is coordinating project updates.`);
    keyPoints.push('Requests input or participation in an upcoming sync-up meeting.');
    keyPoints.push('Highlights deliverables and mentions milestone check-ins.');
  } else if (category === 'Finance') {
    summary = `Billing notification regarding recent statement transaction.`;
    keyPoints.push('Indicates transaction completed or invoice details generated.');
    keyPoints.push('Mentions total amount, order information, and payment methods.');
    keyPoints.push('Contains receipt file instructions for standard tax accounting.');
  } else {
    summary = `General message from ${fromName} regarding '${subject}'.`;
    keyPoints.push('Sender is reaching out for general communications.');
    keyPoints.push('Highlights standard updates and conversational topics.');
    keyPoints.push('Awaiting follow-up response at your convenience.');
  }

  // 4. Task / Deadline extraction
  if (content.includes('by tomorrow') || content.includes('due tomorrow')) {
    const due = new Date();
    due.setDate(due.getDate() + 1);
    tasks.push({
      text: `Complete task mentioned in email: "${subject}"`,
      dueDate: due.toISOString().split('T')[0]
    });
  } else if (content.includes('by friday') || content.includes('due friday')) {
    tasks.push({
      text: `Follow up on action item: "${subject}" before Friday`,
      dueDate: new Date(Date.now() + 3 * 24 * 60 * 60 * 1000).toISOString().split('T')[0]
    });
  } else if (content.includes('meeting at') || content.includes('sync up at')) {
    tasks.push({
      text: `Attend schedule sync-up meeting: "${subject}"`,
      dueDate: new Date().toISOString().split('T')[0]
    });
  }

  return {
    category,
    priority,
    summary,
    keyPoints,
    sentiment,
    tasks,
    phishingAnalysis: {
      riskScore,
      warnings,
      isPhishing,
      explanation
    }
  };
}

export const AiService = {
  async analyzeEmail(subject: string, body: string, fromName: string, fromEmail: string): Promise<IAiAnalysisResult> {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey || apiKey === 'YOUR_GEMINI_API_KEY') {
      // Return simulated NLP results
      return mockAnalyzeEmail(subject, body, fromName, fromEmail);
    }

    const prompt = `
You are SiftMail's advanced AI email intelligence agent.
Analyze the following email and return a strictly structured JSON response.

EMAIL DETAILS:
From: ${fromName} <${fromEmail}>
Subject: ${subject}
Body:
"""
${body}
"""

YOUR TASK:
1. Categorize this email into exactly one of these labels: 'Important', 'Work', 'Personal', 'Promotions', 'Social', 'Spam', 'Finance', 'Updates', 'Newsletters'.
2. Classify its priority: 'High', 'Medium', or 'Low' (High priorities are urgent requests, deadlines, manager messages, critical updates).
3. Create a one-sentence, clear, action-oriented summary of the email under "summary".
4. Extract 3-5 critical, highly descriptive bullet points under "keyPoints".
5. Detect the overall emotional sentiment of the email (e.g. 'Polite / Collaborative', 'Urgent / Alert', 'Conversational', 'Frustrated') under "sentiment".
6. Extract any calendar tasks, meetings, due dates, action items, or deadlines mentioned. Represent each task as an object under "tasks" containing:
   - "text": The descriptive action item.
   - "dueDate": (Optional) The date in YYYY-MM-DD format (infer relative dates based on current time: 2026-05-23T18:46:48+05:30).
7. Perform a rigorous spam & phishing assessment. Render a JSON object under "phishingAnalysis" containing:
   - "riskScore": A number from 0 to 100 indicating the safety threat (0 is totally safe, 100 is highly malicious).
   - "warnings": A string array listing security flags (e.g., "Demands urgent login credentials", "Suspicious link structure", "Spoofed domain").
   - "isPhishing": A boolean indicating if the email is a fraudulent attempt to steal data/phish or highly suspicious spam.
   - "explanation": AI security analyst explanation of why this risk score was given.

RETURN STRUCTURE:
Ensure your response is ONLY a valid JSON string matching the following structure:
{
  "category": "Work",
  "priority": "High",
  "summary": "Detailed summary line",
  "keyPoints": ["point 1", "point 2"],
  "sentiment": "Professional",
  "tasks": [
    { "text": "Task description", "dueDate": "YYYY-MM-DD" }
  ],
  "phishingAnalysis": {
    "riskScore": 15,
    "warnings": [],
    "isPhishing": false,
    "explanation": "Brief rationale"
  }
}
`;

    try {
      const responseText = await callGemini(prompt, true);
      // Strip markdown code block wrappers if Gemini outputs them
      const cleanJson = responseText.replace(/```json/g, '').replace(/```/g, '').trim();
      return JSON.parse(cleanJson) as IAiAnalysisResult;
    } catch (err) {
      console.warn('⚡ Gemini analysis failed, reverting to Mock NLP Sorter:', err);
      return mockAnalyzeEmail(subject, body, fromName, fromEmail);
    }
  },

  async generateReply(originalSubject: string, originalBody: string, senderName: string, tone: 'Professional' | 'Casual' | 'Short' | 'Formal'): Promise<string> {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey || apiKey === 'YOUR_GEMINI_API_KEY') {
      // Mock replies
      if (tone === 'Professional') {
        return `Dear ${senderName},\n\nThank you for your email. I have received your message regarding "${originalSubject}" and am currently reviewing the details. I will provide you with a comprehensive response shortly.\n\nBest regards,\n[Your Name]`;
      } else if (tone === 'Casual') {
        return `Hey ${senderName},\n\nGot your email about "${originalSubject}"! Thanks for reaching out. Let me look into this and I'll get back to you soon. Speak soon!\n\nCheers,\n[Your Name]`;
      } else if (tone === 'Formal') {
        return `Dear Mr./Ms. ${senderName},\n\nI am writing to acknowledge receipt of your correspondence regarding "${originalSubject}". I appreciate you bringing this matter to my attention. I shall investigate this issue further and reply to you in due course.\n\nSincerely yours,\n[Your Name]`;
      } else {
        return `Hi ${senderName}, thanks for the email. Got it! I'm on it and will follow up shortly. - [Your Name]`;
      }
    }

    const prompt = `
You are SiftMail's auto-draft email assistant. Write a high-quality email reply.

ORIGINAL EMAIL:
Sender: ${senderName}
Subject: ${originalSubject}
Body:
"""
${originalBody}
"""

DESIRED REPLY TONE: ${tone}

YOUR TASK:
Generate a contextually appropriate, helpful, and natural reply to the email.
- Do NOT output HTML.
- Ensure the tone matches "${tone}" perfectly.
- Leave placeholders like "[Your Name]" where appropriate.
- Return ONLY the raw body of the drafted reply, no quotes or introductions.
`;

    try {
      return await callGemini(prompt, false);
    } catch (err) {
      console.warn('⚡ Gemini reply generation failed, using mock reply:');
      // Revert to mock reply if API is down
      if (tone === 'Professional') {
        return `Dear ${senderName},\n\nThank you for your email. I have received your message regarding "${originalSubject}" and am currently reviewing the details. I will provide you with a comprehensive response shortly.\n\nBest regards,\n[Your Name]`;
      } else {
        return `Hi ${senderName}, thanks for the email. Got it! I'm on it and will follow up shortly. - [Your Name]`;
      }
    }
  },

  async translateAndSummarize(subject: string, body: string, targetLanguage: string): Promise<{ subject: string; summary: string }> {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey || apiKey === 'YOUR_GEMINI_API_KEY') {
      return {
        subject: `[Translated to ${targetLanguage}] ${subject}`,
        summary: `This is a translated summary of '${subject}' in ${targetLanguage}. The sender is requesting standard communications. Please review details.`
      };
    }

    const prompt = `
Translate and summarize the following email.
Target Language: ${targetLanguage}

EMAIL CONTENT:
Subject: ${subject}
Body:
${body}

YOUR TASK:
1. Translate the Subject line into ${targetLanguage}.
2. Create a concise 2-sentence summary of the email body in ${targetLanguage}.

Return a JSON object matching this structure:
{
  "subject": "Translated subject line",
  "summary": "Translated summary body"
}
`;

    try {
      const responseText = await callGemini(prompt, true);
      const cleanJson = responseText.replace(/```json/g, '').replace(/```/g, '').trim();
      return JSON.parse(cleanJson);
    } catch (err) {
      return {
        subject: `[Translated] ${subject}`,
        summary: `Standard email summary translated to ${targetLanguage} successfully.`
      };
    }
  }
};

File: backend/app/services/ai_service.py
import os
import json
import re
from typing import Optional, List
from app.core.config import settings
from app.schemas.ai import AIAnalysisSchema, UrgencyLevel, SuggestedReply, ExtractedTaskSchema

class AIService:
    @staticmethod
    async def analyze_email(subject: str, body: str, sender: str) -> AIAnalysisSchema:
        """
        Sift through email content using Gemini 2.5 Flash if API key configured,
        or fallback to native Python NLP heuristic analyzer.
        """
        api_key = settings.GEMINI_API_KEY or os.getenv("GEMINI_API_KEY", "")
        if api_key and api_key.strip() and api_key != "your_gemini_api_key_here":
            try:
                from google import genai
                client = genai.Client(api_key=api_key)
                prompt = f"""You are SiftMail AI, an executive email intelligence engine.
Analyze this email thoroughly and return a valid JSON object strictly matching this schema:
{{
  "urgency": "low" | "medium" | "high" | "critical",
  "urgencyScore": integer (0 to 100),
  "sentiment": "positive" | "neutral" | "urgent" | "frustrated" | "inquiry",
  "summary": "1-2 sentence executive TL;DR summary",
  "keyTakeaways": ["key takeaway 1", "key takeaway 2", ...],
  "suggestedReplies": [
    {{"tone": "Professional & Affirmative", "text": "draft reply..."}},
    {{"tone": "Concise", "text": "draft reply..."}}
  ],
  "extractedTasks": [
    {{"title": "action item", "dueDate": "optional deadline", "priority": "low" | "medium" | "high" | "urgent"}}
  ]
}}

Sender: {sender}
Subject: {subject}
Body:
{body}
"""
                response = client.models.generate_content(
                    model="gemini-2.5-flash",
                    contents=prompt,
                    config={"response_mime_type": "application/json"}
                )
                if response.text:
                    data = json.loads(response.text)
                    return AIAnalysisSchema.model_validate(data)
            except Exception as e:
                print(f"[AIService] Gemini API error: {e}, using heuristic fallback.")

        return AIService._heuristic_analyze(subject, body, sender)

    @staticmethod
    async def generate_reply(subject: str, body: str, sender: str, tone: str = "Professional", custom_notes: Optional[str] = None) -> str:
        """
        Generate contextual smart reply using Gemini 2.5 Flash or heuristic templates.
        """
        api_key = settings.GEMINI_API_KEY or os.getenv("GEMINI_API_KEY", "")
        if api_key and api_key.strip() and api_key != "your_gemini_api_key_here":
            try:
                from google import genai
                client = genai.Client(api_key=api_key)
                prompt = f"""You are SiftMail AI. Draft a high-quality email response to the following email.
Target Tone: {tone}
{f'Custom User Guidance: {custom_notes}' if custom_notes else ''}

Original Sender: {sender}
Subject: {subject}
Original Body:
{body}

Output only the formatted response text in markdown without conversational pleasantries."""
                response = client.models.generate_content(
                    model="gemini-2.5-flash",
                    contents=prompt
                )
                if response.text:
                    return response.text.strip()
            except Exception as e:
                print(f"[AIService] Gemini reply error: {e}, using heuristic draft.")

        return AIService._heuristic_reply(subject, body, sender, tone, custom_notes)

    @staticmethod
    def _heuristic_analyze(subject: str, body: str, sender: str) -> AIAnalysisSchema:
        combined = f"{subject} {body}".lower()

        # Urgency detection
        is_critical = any(kw in combined for kw in ["urgent", "asap", "eod", "immediately", "deadline", "emergency", "critical", "board meeting"])
        is_high = any(kw in combined for kw in ["today", "important", "priority", "review", "sign-off", "pitch", "investor"])
        is_security = any(kw in combined for kw in ["vulnerability", "security", "alert", "cve", "breach"])

        if is_critical:
            urgency = UrgencyLevel.CRITICAL
            urgency_score = 94
        elif is_high or is_security:
            urgency = UrgencyLevel.HIGH
            urgency_score = 82
        elif any(kw in combined for kw in ["fyi", "newsletter", "digest", "roundup", "weekly"]):
            urgency = UrgencyLevel.LOW
            urgency_score = 25
        else:
            urgency = UrgencyLevel.MEDIUM
            urgency_score = 60

        # Sentiment detection
        if any(kw in combined for kw in ["delighted", "impressed", "great", "excellent", "solid", "congrats"]):
            sentiment = "positive"
        elif is_critical:
            sentiment = "urgent"
        elif any(kw in combined for kw in ["delay", "issue", "blocker", "error", "failed", "discrepancy"]):
            sentiment = "frustrated"
        elif "?" in body:
            sentiment = "inquiry"
        else:
            sentiment = "neutral"

        # Action line extraction
        lines = [l.strip() for l in body.split("\n") if l.strip()]
        action_lines = []
        for line in lines:
            if re.match(r"^(\d+\.|\*|-|could you|please|we need|let's|verify|update|send)", line, re.IGNORECASE):
                cleaned = re.sub(r"^(\d+\.|\*|-)\s*", "", line)
                if len(cleaned) > 10:
                    action_lines.append(cleaned)

        key_takeaways = action_lines[:4]
        if not key_takeaways:
            key_takeaways = [
                f"Subject Focus: {subject}",
                f"From: {sender}",
                "Review message details for pending context."
            ]

        extracted_tasks: List[ExtractedTaskSchema] = []
        for item in key_takeaways[:3]:
            extracted_tasks.append(ExtractedTaskSchema(
                title=item if len(item) <= 90 else item[:87] + "...",
                dueDate="Today, EOD" if is_critical else "Upcoming",
                priority="urgent" if urgency == UrgencyLevel.CRITICAL else "high" if urgency == UrgencyLevel.HIGH else "medium"
            ))

        first_name = sender.split()[0] if sender else "there"
        suggested_replies = [
            SuggestedReply(
                tone="Professional & Affirmative",
                text=f"Hi {first_name},\n\nThank you for reaching out regarding \"{subject}\". I have reviewed the requirements and am actively coordinating the deliverables.\n\nBest regards,\nPragyan"
            ),
            SuggestedReply(
                tone="Concise",
                text=f"Hi {first_name}, received and noted. Looking into this right away and will follow up shortly."
            )
        ]

        return AIAnalysisSchema(
            urgency=urgency,
            urgencyScore=urgency_score,
            sentiment=sentiment,
            summary=f"Automated Sift: Email from {sender} concerning '{subject}'. Identified {len(extracted_tasks)} critical discussion points.",
            keyTakeaways=key_takeaways,
            suggestedReplies=suggested_replies,
            extractedTasks=extracted_tasks
        )

    @staticmethod
    def _heuristic_reply(subject: str, body: str, sender: str, tone: str, custom_notes: Optional[str]) -> str:
        first_name = sender.split()[0] if sender else "there"
        tone_lower = tone.lower()

        extra = f"\n\nNote: {custom_notes}" if custom_notes else ""

        if "urgent" in tone_lower or "concise" in tone_lower:
            return f"Hi {first_name},\n\nReceived and noted. I am addressing the key action items for \"{subject}\" right now and will update you shortly.{extra}\n\nBest,\nPragyan"
        elif "friendly" in tone_lower:
            return f"Hey {first_name},\n\nThanks for reaching out! This looks great. I'll review the details regarding \"{subject}\" and get back to you with next steps.{extra}\n\nHave a great week,\nPragyan"
        elif "formal" in tone_lower or "executive" in tone_lower:
            return f"Dear {first_name},\n\nThank you for your correspondence regarding \"{subject}\". I have noted the deliverables and our team is actively moving forward on the timeline.{extra}\n\nSincerely,\nPragyan Paramita"
        else:
            return f"Hi {first_name},\n\nThank you for the update. I have reviewed the details for \"{subject}\" and will follow up with the required deliverables on schedule.{extra}\n\nBest regards,\nPragyan"
